src/nanoarrow/schema.c

```diff
@@ -75,7 +163,144 @@ int ArrowSchemaInit(struct ArrowSchema* schema) {
+  if (schema->format == NULL && data_type != NANOARROW_TYPE_INVALID) {
+    return EINVAL;
+  } else {
+    return NANOARROW_OK;
+  }
+}
+
+ArrowErrorCode ArrowSchemaSetFixedSize(struct ArrowSchema* schema, int32_t fixed_size) {
+  struct ArrowSchemaView schema_view;
+  struct ArrowError error;
+  int result = ArrowSchemaViewInit(&schema_view, schema, &error);
+  if (result != NANOARROW_OK) {
+    return result;
+  }
+
+  char buffer[64];
+  int n_chars;
+  switch (schema_view.storage_data_type) {
+    case NANOARROW_TYPE_FIXED_SIZE_BINARY:
+      n_chars = snprintf(buffer, sizeof(buffer), "w:%d", fixed_size);
+      break;
+    case NANOARROW_TYPE_FIXED_SIZE_LIST:
+      n_chars = snprintf(buffer, sizeof(buffer), "+w:%d", fixed_size);
+      break;
+    default:
+      return EINVAL;
+  }
+
+  buffer[n_chars] = '\0';
+  return ArrowSchemaSetFormat(schema, buffer);
+}
+
+ArrowErrorCode ArrowSchemaSetDecimalPrecisionScale(struct ArrowSchema* schema,
+                                                   int32_t decimal_precision,
+                                                   int32_t decimal_scale) {
+  struct ArrowSchemaView schema_view;
+  struct ArrowError error;
+  int result = ArrowSchemaViewInit(&schema_view, schema, &error);
+  if (result != NANOARROW_OK) {
+    return result;
+  }
+
+  char buffer[64];
+  int n_chars;
+  switch (schema_view.storage_data_type) {
+    case NANOARROW_TYPE_DECIMAL128:
+      n_chars =
+          snprintf(buffer, sizeof(buffer), "d:%d,%d", decimal_precision, decimal_scale);
+      break;
+    case NANOARROW_TYPE_DECIMAL256:
+      n_chars = snprintf(buffer, sizeof(buffer), "d:%d,%d,256", decimal_precision,
+                         decimal_scale);
+      break;
+    default:
+      return EINVAL;
+  }
+
+  buffer[n_chars] = '\0';
+  return ArrowSchemaSetFormat(schema, buffer);
+}
+
+ArrowErrorCode ArrowSchemaSetTimeUnit(struct ArrowSchema* schema,
+                                      enum ArrowTimeUnit time_unit) {
+  struct ArrowSchemaView schema_view;
+  struct ArrowError error;
+  int result = ArrowSchemaViewInit(&schema_view, schema, &error);
+  if (result != NANOARROW_OK) {
+    return result;
+  }
+
+  char time_unit_str[2];
+  time_unit_str[1] = '\0';
+  switch (time_unit) {
+    case NANOARROW_TIME_UNIT_SECOND:
+      time_unit_str[0] = 's';
+      break;
+    case NANOARROW_TIME_UNIT_MILLI:
+      time_unit_str[0] = 'm';
+      break;
+    case NANOARROW_TIME_UNIT_MICRO:
+      time_unit_str[0] = 'u';
+      break;
+    case NANOARROW_TIME_UNIT_NANO:
+      time_unit_str[0] = 'n';
+      break;
+    default:
+      return EINVAL;
+  }
+
+  char buffer[64];
+  int n_chars;
+  switch (schema_view.storage_data_type) {
+    case NANOARROW_TYPE_TIME32:
+    case NANOARROW_TYPE_TIME64:
+      n_chars = snprintf(buffer, sizeof(buffer), "tt%s", time_unit_str);
+      break;
+    case NANOARROW_TYPE_TIMESTAMP:
+      n_chars = snprintf(buffer, sizeof(buffer), "ts%s", time_unit_str);
```

Comment:
this will overwrite any existing timezone, right? do we want to preserve it?

Reply:
I guess what if they set the timezone before the time unit? (Not a big deal I suppose. Just needs to be noted clearly.)

Reply:
Or else, set the timezone and time unit together perhaps? (That might be annoying in the common case, though.)